modules/ml/src/main/java/org/apache/ignite/ml/knn/regression/KNNRegressionModel.java

```diff
@@ -13,93 +13,104 @@
 
     /** {@inheritDoc} */
-    @Override public Double predict(Vector v) {
-        List<LabeledVector> neighbors = findKNearestNeighbors(v);
+    @Override public Double predict(Vector input) {
+        List<LabeledVector<Double>> neighbors = findKClosest(k, input);
 
-        return predictYBasedOn(neighbors, v);
+        return predictor.predict(neighbors, input);
     }
 
-    /** */
-    private double predictYBasedOn(List<LabeledVector> neighbors, Vector v) {
-        switch (stgy) {
-            case SIMPLE:
-                return simpleRegression(neighbors);
-            case WEIGHTED:
-                return weightedRegression(neighbors, v);
-            default:
-                throw new UnsupportedOperationException("Strategy " + stgy.name() + " is not supported");
-        }
+    /**
+     * Util class that makes prediction based on the specified list of nearest neighbours.
+     */
+    private interface KNNRegressionPredictor {
+        /**
+         * Makes a regression prediction based on the specified list of nearest neighbours.
+         *
+         * @param neighbors List of nearest neighbours.
+         * @param pnt Point to calculate distance to.
+         * @return Regression prediction.
+         */
+        public Double predict(List<LabeledVector<Double>> neighbors, Vector pnt);
     }
 
-    /** */
-    private double weightedRegression(List<LabeledVector> neighbors, Vector v) {
-        double sum = 0.0;
-        double div = 0.0;
-        for (LabeledVector<Double> neighbor : neighbors) {
-            double distance = distanceMeasure.compute(v, neighbor.features());
-            sum += neighbor.label() * distance;
-            div += distance;
+    /**
+     * Simple (not weighted) implementation of {@link KNNRegressionPredictor} that makes prediction based on the
+     * specified list of nearest neighbours.
+     */
+    private class KNNRegressionSimplePredictor implements KNNRegressionPredictor {
+        /** {@inheritDoc} */
+        @Override public Double predict(List<LabeledVector<Double>> neighbors, Vector pnt) {
+            if (neighbors.isEmpty())
+                return null;
+
+            double sum = 0.0;
+
+            for (LabeledVector<Double> neighbor : neighbors)
+                sum += neighbor.label();
+
+            return sum / k;
         }
-        if (div == 0.0) // when all neighbours are equal to the given point
-            return simpleRegression(neighbors);
-        return sum / div;
     }
 
-    /** */
-    private double simpleRegression(List<LabeledVector> neighbors) {
-        double sum = 0.0;
-        for (LabeledVector<Double> neighbor : neighbors)
-            sum += neighbor.label();
-        return sum / (double)k;
-    }
+    /**
+     * Weighted implementation of {@link KNNRegressionPredictor} that makes prediction based on the specified list of
+     * nearest neighbours.
+     */
+    private class KNNRegressionWeightedPredictor extends KNNRegressionSimplePredictor {
```

Comment:
What's the reason to create separate class for each case? It can be set up via setting of algorithm. Due to this strategy of splitting of the main algorithm we should split a lot of algorithms in codebase

Reply:
It's not an idea I force to use everywhere. As for me, from a design perspective, this approach looks better than 2 or more method in the main class like `predictSimple`, `predictWeighted`, etc.